src/v/kafka/server/quota_manager.cc

```diff
@@ -147,105 +201,66 @@ ss::future<std::chrono::milliseconds> quota_manager::record_partition_mutations(
-                          * std::chrono::seconds(
-                            *_target_partition_mutation_quota());
-        delay_ms = std::chrono::duration_cast<std::chrono::milliseconds>(rate);
-        std::chrono::milliseconds max_delay_ms(_max_delay());
-        if (delay_ms > max_delay_ms) {
-            vlog(
-              klog.info,
-              "Found partition mutation rate for window of: {}. Client:{}, "
-              "Estimated backpressure delay of {}. Limiting to {} backpressure "
-              "delay",
-              rate,
-              it->first,
-              delay_ms,
-              max_delay_ms);
-            delay_ms = max_delay_ms;
-        }
-    }
-    co_return delay_ms;
-}
+    auto delay = co_await maybe_add_and_retrieve_quota(
+      quota_id, now, [now, mutations](quota_manager::client_quota& cq) {
+          if (!cq.pm_rate.has_value()) {
+              return clock::duration::zero();
+          }
+          auto& pm_rate_tracker = cq.pm_rate.value();
+          auto result
+            = pm_rate_tracker.update_and_calculate_delay<clock::duration>(
+              now, 0);
+          pm_rate_tracker.record(mutations);
+          return result;
+      });
 
-static std::chrono::milliseconds calculate_delay(
-  double rate, uint32_t target_rate, clock::duration window_size) {
-    std::chrono::milliseconds delay_ms(0);
-    if (rate > target_rate) {
-        auto diff = rate - target_rate;
-        double delay = (diff / target_rate)
-                       * static_cast<double>(
-                         std::chrono::duration_cast<std::chrono::milliseconds>(
-                           window_size)
-                           .count());
-        delay_ms = std::chrono::milliseconds(static_cast<uint64_t>(delay));
-    }
-    return delay_ms;
+    co_return duration_cast<std::chrono::milliseconds>(
+      cap_to_max_delay(quota_id, delay));
 }
 
-std::chrono::milliseconds quota_manager::throttle(
-  std::optional<std::string_view> quota_id,
-  uint32_t target_rate,
-  const clock::time_point& now,
-  rate_tracker& rate_tracker) {
-    auto rate = rate_tracker.measure(now);
-    auto delay_ms = calculate_delay(
-      rate, target_rate, rate_tracker.window_size());
-
+clock::duration quota_manager::cap_to_max_delay(
+  std::optional<std::string_view> quota_id, clock::duration delay) {
     std::chrono::milliseconds max_delay_ms(_max_delay());
+    std::chrono::milliseconds delay_ms
+      = std::chrono::duration_cast<std::chrono::milliseconds>(delay);
     if (delay_ms > max_delay_ms) {
         vlog(
           klog.info,
-          "Found data rate for window of: {} bytes. Client:{}, "
+          "Client:{}, "
           "Estimated "
           "backpressure delay of {}. Limiting to {} backpressure delay",
-          rate,
           quota_id,
           delay_ms,
           max_delay_ms);
-        delay_ms = max_delay_ms;
+        delay = max_delay_ms;
     }
-    return delay_ms;
+    return delay;
```

Comment:
I know ultimately that it'll end up as milliseconds, but let's keep precision, just in case.
```suggestion
    clock::duration max_delay = _max_delay();
    if (delay > max_delay) {
        vlog(
          klog.info,
          "Client:{}, "
          "Estimated "
          "backpressure delay of {}. Limiting to {} backpressure delay",
          quota_id,
          std::chrono::duration_cast<std::chrono::milliseconds>(delay),
          std::chrono::duration_cast<std::chrono::milliseconds>(max_delay));
    }
    return std::min(delay, max_delay);
```